### DISCUSSION TITLE
Why NextJS web server listens extra TCP ports?

### DISCUSSION THREAD

**Discussion by vdpvdp** [vercel/next.js#51691] [Help]
### Summary

If i create simple NextJS project with npx create-next-app@latest and run it with "npm run dev" (or build it with "npm run build" and then run it with "npm start") - I will see in Process Explorer - SysInternals there are several node.exe processes. One of them listens 3000 Tcp port:
![image](https://github.com/vercel/next.js/assets/46645075/283e7548-0d0a-4511-bca8-c6d199483804)

and other one listens 2 random ports:
![image](https://github.com/vercel/next.js/assets/46645075/f961656a-1218-43a9-8d0b-0aae0b13992a)

The command line of the second one is "C:\Program Files\nodejs\node.exe" X:\test\node_modules\next\dist\compiled\jest-worker/processChild.js
![image](https://github.com/vercel/next.js/assets/46645075/55897f82-1744-4281-8201-adcd873cfe9c)

Why NextJS listens these extra ports? Can i configure them to be fixed (not random) ?

I have hosting with blocked listening ports. I can open some - but i don't know how to configure it to be fixed.

**Comment by vdpvdp** [comment]
Noone can help me with it ?

**Comment by icyJoseph** [comment]
I think this is part of message passing stablished between the main thread and workers. To be specific, AFAIK, normally, workers and the thread can communicate using IPC, but for some reason, that escapes my knowledge, in some cases, you cannot use that channel.

The alternative is to set up your own channel. Next.js does that to work with the jest-workers, which are not doing testing in this case, but are just a better API to work with workers. 

- https://github.com/vercel/next.js/blob/967b876f6b8a6d7bfaef5749ee915c55ae970306/packages/next/src/server/lib/server-ipc/index.ts#L12

Is one instance where this happens, you can see that this line exists: `ipcServer.listen(0, '0.0.0.0', () => {`, which means, pick a random port: https://nodejs.org/api/net.html#serverlistenport-host-backlog-callback

> If port is omitted or is 0, the operating system will assign an arbitrary unused port, which can be retrieved by using server.address().port after the 'listening'.

**Comment by masad-frost** [comment]
It seems like jest-worker also opens up a couple of ports, no idea what that is.

This is kind of disruptive for automatic port detection by web service providers, especially that they listen on 0.0.0.0. It would make sense to listen on loopback/127.0.0.1 since it seems like these are for local ipc. 

Seems to have started happening last year. @shuding @ijjk any clue?

**Comment by klausi** [comment]
Also got alerted by our Sysadmin team that Nodejs is opening random ports listening on 0.0.0.0 the external interface. We have a firewall, so that is saving our asses here, but this is less than ideal. Nextjs should at least only bind to localhost 127.0.0.1 to not open up ports to the outside.

This also triggers rkhunter when it is looking for malware:
```
Warning: Network TCP port 33369 is being used by /usr/bin/node. Possible rootkit: Volc Rootkit SSH server (divine)
         Use the 'lsof -i' or 'netstat -an' command to check this.
```

I think that Nextjs is a rootkit, which is not good.

It looks like it was resolved in recent Nextjs versions in https://github.com/vercel/next.js/issues/51263 , need to test that.
